subsys/storage/flash_map/flash_map_integrity.c

```diff
@@ -81,50 +71,47 @@ int flash_area_check_int_sha256(const struct flash_area *fa,
 		rc = flash_read(fa->fa_dev, (fa->fa_off + fac->off + pos),
 				fac->rbuf, to_read);
 		if (rc != 0) {
-#if defined(CONFIG_FLASH_AREA_CHECK_INTEGRITY_TC)
-			return rc;
-#else /* CONFIG_FLASH_AREA_CHECK_INTEGRITY_MBEDTLS */
 			goto error;
-#endif
 		}
 
-#if defined(CONFIG_FLASH_AREA_CHECK_INTEGRITY_TC)
-		if (tc_sha256_update(&sha,
-				     fac->rbuf,
-				     to_read) != TC_CRYPTO_SUCCESS) {
-			return -ESRCH;
-		}
+#if defined(CONFIG_FLASH_AREA_CHECK_INTEGRITY_PSA)
+		rc = psa_hash_update(&hash_ctx, fac->rbuf, to_read);
 #else /* CONFIG_FLASH_AREA_CHECK_INTEGRITY_MBEDTLS */
-		if (mbedtls_md_update(&mbed_hash_ctx, fac->rbuf, to_read) != 0) {
+		rc = mbedtls_sha256_update(&hash_ctx, fac->rbuf, to_read);
+#endif
+		if (rc != SUCCESS_VALUE) {
 			rc = -ESRCH;
 			goto error;
 		}
-#endif
 	}
 
-#if defined(CONFIG_FLASH_AREA_CHECK_INTEGRITY_TC)
-	if (tc_sha256_final(hash, &sha) != TC_CRYPTO_SUCCESS) {
-		return -ESRCH;
-	}
+#if defined(CONFIG_FLASH_AREA_CHECK_INTEGRITY_PSA)
+	size_t hash_len;
+
+	rc = psa_hash_finish(&hash_ctx, hash, sizeof(hash), &hash_len);
 #else /* CONFIG_FLASH_AREA_CHECK_INTEGRITY_MBEDTLS */
-	if (mbedtls_md_finish(&mbed_hash_ctx, hash) != 0) {
+	rc = mbedtls_sha256_finish(&hash_ctx, hash);
+#endif
+	if (rc != SUCCESS_VALUE) {
 		rc = -ESRCH;
 		goto error;
 	}
-#endif
 
 	if (memcmp(hash, fac->match, SHA256_DIGEST_SIZE)) {
-#if defined(CONFIG_FLASH_AREA_CHECK_INTEGRITY_TC)
+#if defined(CONFIG_FLASH_AREA_CHECK_INTEGRITY_PSA)
+		/* The operation has already been terminated. */
```

Comment:
Shouldn't this part also `goto error`?
There is `psa_has_abort()` that I think is supposed to be called on failure.

Reply:
No, it doesn't need to. If `psa_hash_finish()` returns successfully there is nothing more to do, and here it is the case (otherwise we will have `goto error` already).
This case is a failure in the flash area check (checksum mismatch), not in the PSA hash operation.